numpy/random/mtrand.pyx

```diff
@@ -3471,7 +3493,11 @@ cdef class RandomState:
             with self.lock, nogil:
                 for i in range(cnt):
                     _dp = (<double*>np.PyArray_MultiIter_DATA(it, 1))[0]
-                    _in = (<long*>np.PyArray_MultiIter_DATA(it, 2))[0]
+                    _in = (<np.npy_intp*>np.PyArray_MultiIter_DATA(it, 2))[0]
+                    if _in > LONG_MAX:
```

Comment:
The check you added here could be implemented using the check_array_constraint above by creating a new constant such as `CONS_NON_NEGATVE_INBOUNDS_LONG`.

Reply:
OK, added a commit for this as `LEGACY_CONS_NON_NEGATVE_INBOUNDS_LONG`.  It is slightly awkward, because the scalar check is typed as `double`.
That is fine in practice since we are only interested in longs that are 32bit (and thus fit double).
(on both sides, I used the knowledge that this only makes sense for integers.)